### PR TITLE
add feature add multi members to list

### PR DESCRIPTION
**Pull request by hairihan**
How can I adding several email addresses into one maillist in one time?
I tried to using following method, but it's only support to add one address.
It will cause a timeout error when there are a plenty amount of address have to add to list.

### CODE REVIEW

**Comment by adamof** [comment]
Can you share what params you feed into the method? Are they in the correct format as here https://documentation.mailgun.com/api-mailinglists.html#examples ?

**Comment by hairihan** [comment]
Hi adamof,

I add following method:

--------

**Add multiple mailing list members (limit 1,000 per call):**

```
curl -s --user 'api:YOUR_API_KEY' \
    https://api.mailgun.net/v3/lists/LIST@YOUR_DOMAIN_NAME/members.json \
    -F upsert=true \
    -F members='[{"address": "Alice <alice@example.com>", "vars": {"age": 26}},{"name": "Bob", "address": "bob@example.com", "vars": {"age": 34}}]'
```

**Sample response:**

```
{
  "message": "Mailing list has been updated",
  "list": {
    "members_count": 7,
    "description": "My updated test mailing list",
    "created_at": "Wed, 06 Mar 2013 11:39:51 GMT",
    "access_level": "readonly",
    "address": "dev@samples.mailgun.org",
    "name": "Test List Updated"
  }
}
```
----
I didn't wrote test for it, coz I'm not familiar with the rspec version that author used in this repo.
But it should be with same format with official docs, I'll check it again and update the Readme.

Best~

**Comment by hairihan** [comment]
@adamof Just updated Readme

**Comment by adamof** [comment]
So does the code you added to the Gem work fine?

**Comment by hairihan** [comment]
@adamof  Yes, it has been used ever since in production ENV of my company's product.

File: README.md [hunk 1]
```diff
@@ -96,6 +96,9 @@ parameters = {
 # Add a member to a list
 @mailgun.list_members.add "devs@your.mailgun.domain", "Q@mi6.co.uk"
 
+# Add multiple mailing list members to a list (limit 1,000 per call)
+@mailgun.list_members.add_multi "devs@your.mailgun.domain", [{"address": "Alice <alice@example.com>", "vars": {"age": 26}}, {"name": "Bob", "address": "bob@example.com", "vars": {"age": 34}}].to_json, {:upsert => true}
```
Comment (adamof) [inline README.md]: Do we have to call `.json` ourselves? Can this be moved inside of `add_multi`?

**Comment by adamof** [comment]
What do you think about this? It saves the users from having to work with json strings and adds a spec.

@HashNuke since I'm touching the `update` method as well, it may introduce some incompatibility. What do you think?

```diff
diff --git a/README.md b/README.md
index eab968c..09af48f 100644
--- a/README.md
+++ b/README.md
@@ -97,10 +97,10 @@ parameters = {
 @mailgun.list_members.add "devs@your.mailgun.domain", "Q@mi6.co.uk"
 
 # Add multiple mailing list members to a list (limit 1,000 per call)
-@mailgun.list_members.add_multi "devs@your.mailgun.domain", [{"address": "Alice <alice@example.com>", "vars": {"age": 26}}, {"name": "Bob", "address": "bob@example.com", "vars": {"age": 34}}].to_json, {:upsert => true}
-    
+@mailgun.list_members.add_multi "devs@your.mailgun.domain", [{"address" => "Alice <alice@example.com>", "vars" => {"age" => 26}}, {"name" => "Bob", "address" => "bob@example.com", "vars" => {"age" => 34}}], {:upsert => true}
+
 # Update a member on a list
-@mailgun.list_members.update "devs@your.mailgun.domain", "Q@mi6.co.uk", "Q", {:gender => 'male'}.to_json, :subscribed => 'no')
+@mailgun.list_members.update "devs@your.mailgun.domain", "Q@mi6.co.uk", "Q", {:gender => 'male'}, :subscribed => 'no')
 
 # Remove a member from a list
 @mailgun.list_members.remove "devs@your.mailgun.domain", "M@mi6.co.uk"
diff --git a/lib/mailgun/list/member.rb b/lib/mailgun/list/member.rb
index fcb4a56..b04218c 100644
--- a/lib/mailgun/list/member.rb
+++ b/lib/mailgun/list/member.rb
@@ -23,13 +23,13 @@ module Mailgun
     # Adds a mailing list member with a given address
     # NOTE Use create instead of add?
     def add(member_address, options={})
-      params = {:address => member_address}
+      params = { :address => member_address }
       Mailgun.submit :post, list_member_url, params.merge(options)
     end
 
     def add_multi(members=[], options={})
-      params = {:members => members}
-      Mailgun.submit :post, list_member_json, params.merge(options)
+      params = { :members => members.to_json }
+      Mailgun.submit :post, list_member_json_url, params.merge(options)
     end
 
     # TODO add spec?
@@ -37,16 +37,15 @@ module Mailgun
 
     # Update a mailing list member with a given address
     def update(member_address, options={})
-      params = {:address => member_address}
+      params = { :address => member_address.to_json }
       Mailgun.submit :put, list_member_url(member_address), params.merge(options)
-    end   
+    end
 
     # Deletes a mailing list member with a given address
     def remove(member_address)
       Mailgun.submit :delete, list_member_url(member_address)
     end
 
-
     private
 
     # Helper method to generate the proper url for Mailgun mailbox API calls
@@ -54,8 +53,8 @@ module Mailgun
       "#{@mailgun.base_url}/lists#{'/' + @address}/members#{'/' + member_address if member_address}"
     end
 
-    def list_member_json
+    def list_member_json_url
       "#{@mailgun.base_url}/lists#{'/' + @address}/members.json"
     end
   end
-end
\ No newline at end of file
+end
diff --git a/spec/list/member_spec.rb b/spec/list/member_spec.rb
index 21265f8..ae974c4 100644
--- a/spec/list/member_spec.rb
+++ b/spec/list/member_spec.rb
@@ -1,56 +1,76 @@
 require 'spec_helper'
 
 describe Mailgun::MailingList::Member do
-
-  before :each do
-    @mailgun = Mailgun({:api_key => "api-key"})		# used to get the default values
-
-    @sample = {
-      :email      => "test@sample.mailgun.org",
-      :list_email => "test_list@sample.mailgun.org",
-      :name       => "test",
-      :domain     => "sample.mailgun.org"
+  let(:mailgun) { Mailgun({api_key: "api-key"}) }
+  let(:samples) do
+    {
+      email:      "testsamples.mailgun.org",
+      list_email: "test_listsamples.mailgun.org",
+      name:       "test",
+      domain:     "sample.mailgun.org"
     }
   end
 
   describe "list members" do
     it "should make a GET request with the right params" do
       sample_response = "{\"items\": [{\"size_bytes\": 0,  \"mailbox\": \"postmaster@bsample.mailgun.org\" }  ]}"
-      mailing_list_members_url = @mailgun.list_members(@sample[:list_email]).send(:list_member_url)
+      mailing_list_members_url = mailgun.list_members(samples[:list_email]).send(:list_member_url)
 
       Mailgun.should_receive(:submit).
         with(:get, mailing_list_members_url, {}).
         and_return(sample_response)
-    
-      @mailgun.list_members(@sample[:list_email]).list
+
+      mailgun.list_members(samples[:list_email]).list
     end
   end
 
   describe "find member in list" do
     it "should make a GET request with correct params to find given email address" do
       sample_response = "{\"items\": [{\"size_bytes\": 0,  \"mailbox\": \"postmaster@bsample.mailgun.org\" }  ]}"
-      mailing_list_members_url = @mailgun.list_members(@sample[:list_email]).send(:list_member_url, @sample[:email])
-      
+      mailing_list_members_url = mailgun.list_members(samples[:list_email]).send(:list_member_url, samples[:email])
+
       Mailgun.should_receive(:submit).
         with(:get, mailing_list_members_url).
         and_return(sample_response)
 
-      @mailgun.list_members(@sample[:list_email]).find(@sample[:email])
+      mailgun.list_members(samples[:list_email]).find(samples[:email])
     end
   end
 
   describe "add member to list" do
     it "should make a POST request with correct params to add a given email address" do
       sample_response = "{\"items\": [{\"size_bytes\": 0,  \"mailbox\": \"postmaster@bsample.mailgun.org\" }  ]}"
-      mailing_list_members_url = @mailgun.list_members(@sample[:list_email]).send(:list_member_url)
+      mailing_list_members_url = mailgun.list_members(samples[:list_email]).send(:list_member_url)
 
       Mailgun.should_receive(:submit).
         with(:post, mailing_list_members_url, {
-          :address => @sample[:email]
+          address: samples[:email]
+        }).
+        and_return(sample_response)
+
+      mailgun.list_members(samples[:list_email]).add(samples[:email])
+    end
+  end
+
+  describe "add multiple members to list" do
+    let(:emails) do
+      [
+        {"address" => "Alice <alice@example.com>", "vars" => {"age" => 26}},
+        {"name" => "Bob", "address" => "bob@example.com", "vars" => {"age" => 34}}
+      ]
+    end
+
+    it "should make a POST request to add the given email addresses" do
+      sample_response = "{\"items\": [{\"size_bytes\": 0,  \"mailbox\": \"postmaster@bsample.mailgun.org\" }  ]}"
+      mailing_list_members_json_url = mailgun.list_members(samples[:list_email]).send(:list_member_json_url)
+
+      Mailgun.should_receive(:submit).
+        with(:post, mailing_list_members_json_url, {
+          members: emails.to_json
         }).
         and_return(sample_response)
 
-      @mailgun.list_members(@sample[:list_email]).add(@sample[:email])
+      mailgun.list_members(samples[:list_email]).add_multi(emails)
     end
   end
 
@@ -58,25 +78,24 @@ describe Mailgun::MailingList::Member do
     it "should make a PUT request with correct params" do
       sample_response = "{\"items\": [{\"size_bytes\": 0,  \"mailbox\": \"postmaster@bsample.mailgun.org\" }  ]}"
       Mailgun.should_receive(:submit).
-        with(:put, "#{@mailgun.list_members(@sample[:list_email]).send(:list_member_url, @sample[:email])}", {
-          :address => @sample[:email]
+        with(:put, "#{mailgun.list_members(samples[:list_email]).send(:list_member_url, samples[:email])}", {
+          address: samples[:email].to_json
         }).
         and_return(sample_response)
 
-      @mailgun.list_members(@sample[:list_email]).update(@sample[:email])
+      mailgun.list_members(samples[:list_email]).update(samples[:email])
     end
   end
 
   describe "delete member from list" do
     it "should make a DELETE request with correct params" do
       sample_response = "{\"items\": [{\"size_bytes\": 0,  \"mailbox\": \"postmaster@bsample.mailgun.org\" }  ]}"
-      mailing_list_members_url = @mailgun.list_members(@sample[:list_email]).send(:list_member_url, @sample[:email])
+      mailing_list_members_url = mailgun.list_members(samples[:list_email]).send(:list_member_url, samples[:email])
       Mailgun.should_receive(:submit).
         with(:delete, mailing_list_members_url).
         and_return(sample_response)
 
-      @mailgun.list_members(@sample[:list_email]).remove(@sample[:email])
+      mailgun.list_members(samples[:list_email]).remove(samples[:email])
     end
   end
-
 end
```

File: README.md (hunk 1 above)
Comment (hairihan) [inline README.md]: Yes, it's better to move "to_json" inside of the method. But does it passed the test?

Comment (hairihan) [inline README.md]: @adamof  I just test it in my local ENV, and it do working well. So I think move inside is OK.

**Comment by HashNuke** [comment]
@adamof These changes are good. They seem to pass tests too.

B/w we need to remove hardcoding responses in tests and instead use https://github.com/vcr/vcr to run tests. I can configure TravicCI with a mailgun API key.

**Comment by HashNuke** [comment]
@adamof @hairihan If you guys are using the gem, I can add you as maintainers. That way I won't be a blocker for improvements.

**Pull request merged**